python/lsst/daf/butler/core/datasets/ref.py

```diff
@@ -0,0 +1,278 @@
+
+    def __eq__(self, other: DatasetRef):
+        return (self.datasetType, self.dataId, self.id) == (other.datasetType, other.dataId, other.id)
+
+    def __hash__(self) -> int:
+        return hash(self.datasetType, self.dataId, self.id)
+
+    def __repr__(self):
+        return f"DatasetRef({self.datasetType}, {self.dataId}, id={self.id}, run={self.run})"
+
+    @property
+    def hash(self) -> bytes:
+        """Secure hash of the `DatasetType` name and data ID (`bytes`).
+        """
+        if not hasattr(self, "_hash"):
+            message = hashlib.blake2b(digest_size=32)
+            message.update(self.datasetType.name.encode("utf8"))
+            self.dataId.fingerprint(message.update)
+            self._hash = message.digest()
+        return self._hash
+
+    @property
+    def components(self) -> Mapping[str, DatasetRef]:
+        """Named `DatasetRef` components.
+
+        Read-only (but not immutable); update via `Registry.attachComponent()`.
+        """
+        return MappingProxyType(self._components)
+
+    @property
+    def dimensions(self) -> DimensionGraph:
+        """The dimensions associated with the underlying `DatasetType`
+        """
+        return self.datasetType.dimensions
+
+    def __str__(self) -> str:
+        components = ""
+        if self.components:
+            components = ", components=[" + ", ".join(self.components) + "]"
+        return "DatasetRef({}, id={}, dataId={} {})".format(self.datasetType.name,
+                                                            self.id, self.dataId, components)
+
+    def __getnewargs_ex__(self) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        return ((self.datasetType, self.dataId),
+                {"id": self.id, "run": self.run, "components": self._components})
+
+    def resolved(self, id: int, run: Run, components: Optional[Mapping[str, DatasetRef]] = None
+                 ) -> DatasetRef:
+        """Return a new `DatasetRef` with the same data ID and dataset type
+        and the given ID and run.
+
+        Parameters
+        ----------
+        id : `int`
+            The unique integer identifier assigned when the dataset is created.
+        run : `Run`
+            The run this dataset was associated with when it was created.
+        components : `dict`, optional
+            A dictionary mapping component name to a `DatasetRef` for that
+            component.
+
+        Returns
+        -------
+        ref : `DatasetRef`
+            A new `DatasetRef`.
+        """
+        newComponents = self._components.copy()
+        if components:
+            newComponents.update(components)
+        return DatasetRef(datasetType=self.datasetType, dataId=self.dataId,
+                          id=id, run=run, hash=self.hash, components=newComponents, conform=False)
+
+    def unresolved(self) -> DatasetRef:
+        """Return a new `DatasetRef` with the same data ID and dataset type,
+        but no ID, run, or components.
```

Comment:
Why are components dropped rather than including them in their unresolved form?

Reply:
See other discussion on components - I don't think we ever want the components on an unresolved `DatasetRef`, and I didn't want to waste effort making them.